// @onboardmoney/cli/lib/commands/balance.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("@oclif/command");
const cli_ux_1 = require("cli-ux");
const sdk_1 = require("@onboardmoney/sdk");
class Balance extends command_1.Command {
    async run() {
        var _a, _b;
        // parse args
        const { args, flags } = this.parse(Balance);
        // pull apikey from arg or cache
        const apikey = (_a = args.apikey) !== null && _a !== void 0 ? _a : '';
        // validate apikey
        if (!apikey) {
            this.error(`invalid apikey ${apikey}`);
        }
        // pull network from arg or cache
        const network = (_b = args.network) !== null && _b !== void 0 ? _b : '';
        // validate network
        const networks = ['mainnet', 'ropsten', 'kovan', 'rinkeby', 'goerli'];
        if (!networks.includes(network)) {
            this.error(`invalid network: ${network}`);
        }
        // init client
        const onboardmoney = new sdk_1.App(apikey, `https://${network}.onboard.money`);
        // query balance
        cli_ux_1.default.action.start('Querying balance');
        const { balance, relayAddress } = await onboardmoney.balance();
        cli_ux_1.default.action.stop(`Relay on ${network} at ${relayAddress} has ${balance} ETH`);
        cli_ux_1.default.url(relayAddress, network === 'mainnet'
            ? `https://etherscan.io/address/${relayAddress}`
            : `https://${network}.etherscan.io/address/${relayAddress}`);
    }
}
exports.default = Balance;
Balance.description = 'Fetch application ETH balance and address for gas payments';
Balance.examples = [
    `$ onboardmoney balance --network=mainnet
Querying balance...
Relay at 0x1234 has 0.1 ETH
https://etherscan.io/address/0x1234
`,
];
Balance.flags = {
    help: command_1.flags.help({ char: 'h' }),
    network: command_1.flags.string({
        char: 'n',
        description: 'network to target',
    }),
    apikey: command_1.flags.string({
        char: 'a',
        description: 'apikey to attach',
    }),
};
Balance.args = [{ name: 'network' }, { name: 'apikey' }];
